support/cas-server-support-trusted-mfa/src/main/java/org/apereo/cas/trusted/config/MultifactorAuthnTrustedDeviceFingerprintConfiguration.java

```diff
@@ -0,0 +1,153 @@
+
+import static org.apereo.cas.trusted.BeanNames.BEAN_DEVICE_FINGERPRINT_COOKIE_CIPHER_EXECUTOR;
+import static org.apereo.cas.trusted.BeanNames.BEAN_DEVICE_FINGERPRINT_COOKIE_GENERATOR;
+import static org.apereo.cas.trusted.BeanNames.BEAN_DEVICE_FINGERPRINT_COOKIE_RANDOM_STRING_GENERATOR;
+import static org.apereo.cas.trusted.BeanNames.BEAN_DEVICE_FINGERPRINT_STRATEGY;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.apereo.cas.CipherExecutor;
+import org.apereo.cas.configuration.CasConfigurationProperties;
+import org.apereo.cas.configuration.model.core.util.EncryptionJwtSigningJwtCryptographyProperties;
+import org.apereo.cas.configuration.model.support.mfa.trusteddevice.DeviceFingerprintProperties;
+import org.apereo.cas.configuration.model.support.mfa.trusteddevice.DeviceFingerprintProperties.ClientIp;
+import org.apereo.cas.configuration.model.support.mfa.trusteddevice.DeviceFingerprintProperties.Cookie;
+import org.apereo.cas.configuration.model.support.mfa.trusteddevice.DeviceFingerprintProperties.UserAgent;
+import org.apereo.cas.trusted.util.cipher.CookieDeviceFingerprintComponentCipherExecutor;
+import org.apereo.cas.trusted.web.flow.ClientIpDeviceFingerprintComponent;
+import org.apereo.cas.trusted.web.flow.CookieDeviceFingerprintComponent;
+import org.apereo.cas.trusted.web.flow.DefaultDeviceFingerprintStrategy;
+import org.apereo.cas.trusted.web.flow.DeviceFingerprintComponent;
+import org.apereo.cas.trusted.web.flow.DeviceFingerprintStrategy;
+import org.apereo.cas.trusted.web.flow.UserAgentDeviceFingerprintComponent;
+import org.apereo.cas.trusted.web.support.TrustedDeviceCookieRetrievingCookieGenerator;
+import org.apereo.cas.util.gen.Base64RandomStringGenerator;
+import org.apereo.cas.util.gen.RandomStringGenerator;
+import org.apereo.cas.web.support.CookieRetrievingCookieGenerator;
+import org.apereo.cas.web.support.CookieValueManager;
+import org.apereo.cas.web.support.EncryptedCookieValueManager;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.context.config.annotation.RefreshScope;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+import java.util.List;
+
+/**
+ * Configuration for {@link DefaultDeviceFingerprintStrategy}.
+ *
+ * @author Daniel Frett
+ * @since 5.3.0
+ */
+@Slf4j
+@Configuration("multifactorAuthnTrustedDeviceFingerprintConfiguration")
+@EnableConfigurationProperties(CasConfigurationProperties.class)
+public class MultifactorAuthnTrustedDeviceFingerprintConfiguration {
+    @Autowired
+    private CasConfigurationProperties casProperties;
+
+    @Bean
+    @RefreshScope
+    public DeviceFingerprintComponent deviceFingerprintClientIdComponent() {
```

Comment:
You should try to rework the design so that beans are only generated if `enabled` is true. Bean creation must be made conditional, and then you can ask the application context for all components that are created and register them appropriately.

Reply:
I'm not sure if `@ConditionalOnProperty` works with `@RefreshScope` based on my understanding of how `@RefreshScope` is implemented.

RefreshScoped beans create a proxy to the actual bean, that proxy is wired into other components that require the bean. With `@ConditionalOnProperty` the bean (and proxy) wouldn't get created if `@ConditionalOnProperty` disables the bean. At this point it would not get wired in elsewhere and changing the property to enable the bean wouldn't actually enable the bean and update wiring.

With that in mind I opted to use a NoOp fingerprint component instead so that the component can be enabled/disabled and still be refreshed within the active spring context.

Reply:
Ah, I see. I suppose using RefreshScope something that one could successfully take advantage of, in this case. Thanks for the note. LGTM.